test/Microsoft.ML.TestFrameworkCommon/Attributes/VaryingToleranceAttribute.cs

```diff
@@ -0,0 +1,37 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using Xunit.Sdk;
+namespace Microsoft.ML.TestFrameworkCommon.Attributes
+{
+    /// <summary>
+    /// A theory for tests with varying tolerance levels
+    /// <code>
+    /// [Theory, VaryingTolerance(5)]
+    /// public void VaryingToleranceTest(double tolerance)
+    /// {
+    ///     Console.WriteLine(String.Format("Tolerance: {0}", tolerance)); // "Tolerance: 1e-5
+    ///     ...
+    ///     Assert.InRange(actualMetric, expectedMetric - tolerance, expectedMetric + tolerance);
+    /// }
+    /// </code>
+    /// </summary>
+    public sealed class VaryingToleranceAttribute : DataAttribute
+    {
+        public VaryingToleranceAttribute(int tolerance)
+        {
+            Tolerance = tolerance;
+        }
+
+        public int Tolerance { get; }
+        public override IEnumerable<object[]> GetData(MethodInfo testMethod)
+        {
+            Console.WriteLine(String.Format("Test \"{0}\" utilizes varying tolerances.", testMethod.Name));
+            yield return new object[] { Tolerance };
```

Comment:
Why yield? Why not directly return an enumerable?

Reply:
Hi Harish. I utilized [this](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/yield)documentation in my original implementation and the pre-existing usage of [GetData in IterationDataAttribute](https://github.com/dotnet/machinelearning/blob/4493397fc35060d7d04c9333bf00164d320a6621/test/Microsoft.ML.TestFrameworkCommon/Attributes/IterationDataAttribute.cs#L29-L36), but now I realize with your review thatI'm not iterating when calling GetData, I can just return `(IEnumerable<object[]>)new object[] { Tolerance }`. Thank you!

Reply:
Hi Harish. I utilized [this](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/yield) documentation in my original implementation and the pre-existing usage of [GetData in IterationDataAttribute](https://github.com/dotnet/machinelearning/blob/4493397fc35060d7d04c9333bf00164d320a6621/test/Microsoft.ML.TestFrameworkCommon/Attributes/IterationDataAttribute.cs#L29-L36). I attempted to return `(IEnumerable<object[]>)new object[] { Tolerance }` instead of using `yield`, but I get the following exception:

>     System.InvalidCastException : Unable to cast object of type 'System.Collections.Generic.List`1[System.Int32]' to type 'System.Collections.Generic.IEnumerable`1[System.Object[]]'.

By using `yield` I am able to bypass this restriction. Is there something I'm overlooking here?